GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -288,15 +318,66 @@ public void Add(GitRevision revision, int? insertScore = null, int insertRange =
                 }
             }
 
-            // Ensure all parents are loaded before adding it to the _nodes list. This is important for ordering.
-            ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), revisionGraphRevision);
+            _revisionByObjectId.TryAdd(revision.ObjectId, revisionGraphRevision);
+            _revisionCount++;
+            return MarkCacheAsInvalidIfNeeded(minExistingScore);
+        }
 
-            if (!updateParents)
+        /// <summary>
+        /// Insert workTree and index revisions to the graph, offset existing revisions.
+        /// </summary>
+        /// <param name="workTreeRev">The workTree revision to add.</param>
+        /// <param name="indexRev">The index revision to add.</param>
+        /// <param name="parents">Parent ids for the revision to find (and insert before).</param>
+        /// returns <see langword="true"/> if the cache is invalidated; otherwise <see langword="false"/>.
+        public bool Insert(GitRevision workTreeRev, GitRevision indexRev, IReadOnlyList<ObjectId> parents)
+        {
+            DebugHelpers.Assert(!_revisionByObjectId.ContainsKey(workTreeRev.ObjectId), "Worktree revision was unexpectedly already added to the RevisionGraph.");
+            DebugHelpers.Assert(!_revisionByObjectId.ContainsKey(indexRev.ObjectId), "Index revision was unexpectedly already added to the RevisionGraph.");
```

Comment:
It may be clearer here. But why search again in addition to `TryAdd`.

Reply:
This is only for Debug, optimized in Release
Not done for Add, the assert was adding a lot of time (like 100ms for the linux 100K repo)

Reply:
I mean `.ContainsKey` does the same search a second time as `.TryAdd` needs to do again below. This slows down the debug build.
So just verify the return value from `.TryAdd` using `DebugHelpers.Assert`.

Reply:
I understood. This is not affecting performance as it is only called once when HEAD is not in the commits.
A similar check in Add() had surprisingly much impact.
If the check is combined there is an unused variable in Release
```
            bool bw = _revisionByObjectId.TryAdd(workTreeRev.ObjectId, workTreeGraphRevision);
            DebugHelpers.Assert(bw, "Worktree revision was unexpectedly already added to the RevisionGraph.");
```